[MDB Ignore] Unrestricted Airlocks - Door Delays

san7890 opened the pull request:
## About The Pull Request

This PR adds a change to a pretty prevalent feature that's been around since #66588 (April 28th, 2022). Although it has been controversial to some extent, I believe that it was a good change in principle for reasons I will later get into. For now, this is what the PR does:

* All _station_ unrestricted door helpers (e.g. those in maintenance) have been replaced with a subtype that adds a random 2-3 second delay to opening the door, using a new "unrestricted latch" feature. **This does not occur if the user has regular access to that door.**
* The unrestricted side will initiate a do_after and plays a distinct small hissing sound (because I presume you're disengaging some hydraulic lever on that side), after which the door will open as players have come to expect.
* While you're engaging the lever, you will be resistant to pressure movements as you are presumably holding onto the lever or something. All other movements (e.g. voluntary, being shoved, etc.) will continue to function as expected.
* We will statistically tabulate each time someone attempts to use the unrestricted "lever", and every time one successfully occurs. This is just for analysis to see player trends as we have lacked this over the last 3.5 years.

Here's a link to the sound I added: https://freesound.org/people/Usernameis1337/sounds/632755, the version in this PR was converted using OGG Vorbis.

https://github.com/user-attachments/assets/4167bda4-63b5-4dc0-a7b2-694a9e3d3201

## Why It's Good For The Game

Back in 2022, I was one of the major proponents for adding such a change to the game. This is for a few reasons:

* It doesn't really make human sense for a station to have hallways/spaces that you can enter and then just get completely trapped in. The station is a deathtrap because of _who_ is on it, not because it _is_ a death trap. I have always maintained that we should assume that the station is competently made to give the hijinx a much more jarring background. To me, maintenance isn't old sections of a station (although it can have those elements), but vital side pathways in the event of emergencies, like having extra fire stairs in a building. The doors to those fire stairs are always unlocked because they need to ensure that people are able to efflux to save themselves from danger rather than die.
* It was easily subvertable anyways. You can easily ask the HoP for maintenance access or AI to let you out of maintenance (although the latter was a bit harder to articulate). I don't find myself agreeing with this as much nowadays but I remember the frustration I felt waiting for minutes for someone to open the door had something occurred for me to have found myself in maintenance. It's simply not fun and doesn't make sense for a station to have this.
* I thought that it added a unique element to antagonism. You can't drag someone in maintenance and then have the whole entire subsection of the station be your den. It's important for you to secure your base should you be dragging people in there instead of just always being able to win via dragging.
* People tend to forget this point, but it actually did cause a lot _more_ deaths in situations where threats originate outside the station/in maintenance. It was much easier for the killy-stabby mass murderers to ingress into main hallways and get to killing faster. This is why the change felt "equal" in a way.

### Why not remove it outright?

I believe that a lot of balance changes over the last 3.5 years (30000 issues ago!) have been weighting for these interactions, and I think removing it all cold-turkey like will cause a huge upturn in deadliness and just not be what players are looking for in the new game loop. To me, this was a vital step in getting more towards 90-minute rounds, and I think we would see a drop in that. I also have those above three reasons that I am still a firm believer in, but I am still interested in seeing what a different system might entail and how that could play out. I added stat gathering so after we get 4-6 weeks of data, we can determine to keep or shelve this feature. I will probably always be on the "blocking" side of removing it outright unless there's some severe data that can convince me (which will probably need it's own test-merged PR, while this is meant to be added into the game right now).

### Why are we changing it?

I think not enough people are dying. I played a few rounds yesterday and while a decent chunk of people were dying, I saw a few follies in my original line of thought. Maintenance should be a bit creepier and not auxiliary hallways (in normal circumstances), I already thought we did a really good job of that but I noticed that it wasn't as good as memory served me. I think maintenance should have a little bit more difficulty to navigate (i.e. if you're being chased down by someone and it's taking 2.5 seconds to open, that would be an excellent heart-pounding tension moment) and would overall be more interesting than the status quo. It is a bit sillier to me but some IRL buildings do have to have you hold down the push bar for a few seconds until it disengages, so this is agreeable.

I also don't think the unidirectional exit design flow standards are where they were due to drift in application, so this is also a better way to uniformly apply it.
## Changelog
:cl:
balance: After noticing it was too easy for animals to crawl through maintenance by abusing the Easy-Exit sensors, Nanotrasen has replaced all unrestricted door sensors (like in maintenance) with a pull-down lever that should take around 2-3 seconds to engage. It has been designed to allow the employee to actuate it despite any pressure differentials within an area.
balance: The above change does not apply to select doors in the medical section of the station, which will continue to use the "older" Ready-Exit system.
sound: In order to ensure that staff in high-traffic areas don't go bonkers due to this new change, sound dampeners have been installed to prevent the hydraulics from being heard for super long distances. However, they skimped out on the quality for the ones in maintenance...
/:cl:

FYI Mappers/Tweakers: This PR replaces all station map unres mapping helpers with the new subtype. You are welcome to replace certain doors with the old helper (the base type) to get the old behavior as you see fit, just explain why that behavior is desired in that location. Otherwise, it would be appreciated if we have this new behavior >95% of the time.

tontyGH commented:
Interesting idea. I'd like to see this explored with possibly unrestricted maints, since it seems like it would be a pretty effective deterrent from prowling (maybe make the opening sound go through walls?)

Anyways, let's see what the Terry players think.

san7890 committed:
Directional Delay - Code

san7890 committed:
Directional Delay - Maps

san7890 committed:
sensor -> latch

makes me feel better

san7890 committed:
oops i didn't overload the subtype correctly

san7890 committed:
culls incapacitated multiplier

doesn't actually do anything whoops

san7890 committed:
Revert "Directional Delay - Maps"

This reverts commit cfcbe749a3a707166b87bcf7b35826485c2486c0.

san7890 committed:
better update paths

san7890 committed:
maps for real this time

san7890 committed:
whoops forgot the icon states

san7890 committed:
rename vars

san7890 committed:
explicitly play sound on turf

mikederkan commented:
Does this affect doors in departments like in Medical, with the one-way exit for non-medical staff?

san7890 commented:
> I think there should be a subtype without the delay specifically for high-traffic areas like medbay, where it's meant as a QOL feature more than as a safety precaution.

This already exists, it's the parent type (green) that still exists on non-station maps. Only the subtype of that type (colored yellow) is what's adding the delay feature to doors. I am okay with people coming back to add unrestricted nondelayed helpers, but I just want to see if it's even necessary to begin with using some data collection (a week's worth should be okay).

I'm going to add a line to the statistical telemetry that logs what doortypes see the usage so I can better determine the direction I want to take with the feature, just want to get the data now and have it be done with rather than have regrets down the line.

san7890 committed:
adds types to tally message

works on my heidisql

Layzu666 commented:
I like this idea, but 2.5 seconds to open the door is too much.

glue0000 commented:
retched PR 

criticism is usually pointless (this will be merged - it just will, we know it)
i ask myself the fuck do I even bother

antags have to utilize maintenance to survive (dead ends are a death sentence).

the being cuckblocked by a needless frustrating restriction (we seem to love those nowadays) while being chased by sec, forced to stand still and hug a door for up to 3 seconds. the guess work knowing which door is going will cuck you. the abhorrent maint setups ((CERTAIN MAPS)) have.

> if you're being chased down by someone and it's taking 2.5 seconds to open, that would be an excellent heart-pounding tension moment

heart pounding tension is usually Sec vs Antag keeping an ongoing chase. can be absolutely thrilling stuff to both observe and be apart of. 
cuckblocked by a 3 second door hugging session to be hit a winstick - because dead ends for antags are almost always means a death sentence.

> I think not enough people are dying.

do you play Manuel? dynamic hasn't been touched by the head-admins for 4-5 months. no shit people aren't dying man. it ain't your fault. this doesn't get solved by restricting maints more.

san7890 committed:
adds temporary pressure resistance

san7890 committed:
make it compile

san7890 committed:
whoops i forgot the important part

san7890 committed:
okay i guess we need a CD

san7890 commented:
https://github.com/user-attachments/assets/3be11aa0-acc7-433c-916d-21853fe6b49e

I added it, here's how it looks.

san7890 committed:
makes the hissing sound less annoying

adds sound dampening in non-maintenance areas and does some different trickery in maintenance areas. should be less annoying in high traffic places i.e. medbay. i think the hiss should be synonymous with latch operation but we can nerf it even further if annoying.

san7890 committed:
i vibe coded a bit too hard + fixes an issue i made

san7890 committed:
Revert "maps for real this time"

This reverts commit 2874a0de69a4a78a754673c19e52318e96306c5a.

san7890 committed:
Merge remote-tracking branch 'upstream/master' into directional-delay

san7890 committed:
reapply map changes

san7890 committed:
Revert "reapply map changes"

This reverts commit 3364626230bd735c5ada50536a75b805034fab0c.

san7890 committed:
Merge remote-tracking branch 'upstream/master' into directional-delay

san7890 committed:
Re-run UpdatePaths

san7890 committed:
removes the delay from medbay doors

san7890 commented:
As of the latest commit, the restriction from medbay doors have been removed. The PR was never updated to have the sound changes but I don't think it had any meaningful impact in gameplay from what I've been hearing anyways.

archbtw0 commented:
sad
